### ISSUE TITLE
Can't run a subprocess from within a `py_image`

### ISSUE THREAD

**Issue by joshclimacell** [bazelbuild/rules_docker#341]
TL;DR: Can't run data dependencies as subprocesses within a `py_image`...

Hello!

I have two Python binaries configured to build with Bazel, call them `sub.py` and `bin.py`. I want to run `sub` in a subprocess from within `bin`. (I need to do this because one is Python 2-only while the other is Python 3-only, so I can't get away with converting `sub` to a library. Python sucks...) This works no problem with the `subprocess` module if I make `sub` a `par_binary` (using subpar) and `bin` a `py_binary`; e.g.:

```python
par_binary(
    name = "sub",
    ...
)

py_binary(
    name = "bin",
    data = [":sub"],
    ...
)
```

How do I get this to work with Docker, though? When I try making `bin` into a `py_image`, running the subprocess fails with:

```
OSError: [Errno 2] No such file or directory
```

even though the file definitely exists (e.g. `os.path.exists(<path>)` returns `True`, and manually checking shows permissions/chmod is set up correctly). Indeed, if I attach to the container (built with `-c dbg`) and try to run `./sub.par` directly, I get

```
sh: /app/bin.binary.runfiles/__main__/bin.par: not found
```

which I don't understand since `ls` says it does exist. Any help would be appreciated!

**Comment by joshclimacell** [comment]
(Or really I'm wondering more generally what the preferred method is of packaging other binaries to be run as subprocesses... Seems like a pretty important thing to be able to do?)

**Comment by mattmoor** [comment]
I wonder if the `.par` file has a `#!` line that isn't found?  What happens when you explicitly prefix it with `/usr/bin/python`?

**Comment by mattmoor** [comment]
@duggelz FYI

**Comment by joshclimacell** [comment]
Thanks for the response Matt! Yes, making it `subprocess.call(['python', <path to par>])` does work (well, it doesn't, because wrong version of Python, but that's a separate issue...), but for example if I change `sub` above to a `cc_binary` instead, then subprocess call it from a `py_image`, I get a similar error:

```python
OSError: [Errno 8] Exec format error
```

Which sounds like the base image doesn't have all the right OS libraries or something... I guess maybe I need to have my own special base image or something?

**Comment by mattmoor** [comment]
If you want to juggle tools, then you probably want a more complete base.  These base images are optimized for applications that favor library calls over shelling out to commands (even your own).

I am a little surprised `cc_binary` wouldn't work since the `py_image` base builds on the `cc_image` base.  Perhaps this is a cross-compilation issue, are you building from linux?

What if you use `go_binary` with `pure = "on"` and `--platforms=@io_bazel_rules_go//go/toolchain:linux_amd64`?  IDK the cross-compilation voodoo for C++ off the top of my head :)

**Comment by joshclimacell** [comment]
Ah, no, duh, I'm building on a Mac. That's probably the issue! Thanks! Maybe there should be a restriction somewhere in the rules about embedding e.g. a `cc_binary` built on Mac/Windows/etc. into one of these Linux images? But doesn't seem high priority so closing this now...

**Comment by mattmoor** [comment]
@joshclimacell Yeah, I'd love a portable (across languages) way of telling Bazel that the files we get should be compatible with a particular platform/architecture.  However, in theory we could start building windows/arm containers with this setup, so I'd settle for a portable way of specifying the compilation target :)

**Comment by mattmoor** [comment]
@davidstanke FYI
